[New concept + exercise]: Case (#1595)
exercism/ruby: concepts/case/about.md

# Case

> [Case][case] (often referred to as switch in other languages) is a form of control expression like if-else.

Reviewer: ```suggestion
[Case][case] (often referred to as switch in other languages) is a form of flow control expression like if-else.
```

> Case allows for chaining of multiple if-else-if statements and can be more readable and also allows for powerful constructs.

Reviewer: ```suggestion
Case allows for chaining of multiple `when` statements and can be more readable and also allows for powerful constructs.
```

Also, does it really allow for "more powerful constructs"?  Or are they just as powerful but different?

Author: It doesnt really say they are more powerful just that they allow for powerful constructs.
And it is suppose to be if-else-if.

Reviewer: I can see what it says, if "it" is this copy.  But should it?

Reviewer: And, yes, while we can indeed use if-else-end along and even inside of case statements, the case statement does not itself use that syntax.

Reviewer: I would even say that it is allowed for, even without case.  So the case statement isn't what is allowing for chaining if-else-if statements.

> Case allows for chaining of multiple if-else-if statements and can be more readable and also allows for powerful constructs.

Reviewer: ```suggestion
Case allows for chaining of multiple if-else-if statements and can be more readable while still providing flow control.
```

How about that?

A case is defined by the keyword `case` followed by an optional expression.
Then for each case, the keyword `when` is used followed by an expression which is compared to the case expression.
The `when` keyword should not be indented from the `case` keyword.
After the `when` keyword is the code that should be executed if the case expression matches the when expression.

> Case allows for an optional `else` statement which is executed if no other case matches.

Reviewer: ```suggestion
Case allows for an optional `else` statement which is executed if no other `when` matches.
```

The case expression is evaluated and then compared to each `when` expression.

> The expression is compared using the case subsumption operator (`===`).

Reviewer: In Ruby, we call that the "case equality operator".

Reviewer: I have seen Jorg (from stack overflow) use it in 2010 and someone else call it this in their blog in 2022, but this is only the third time I have seen this `===` called that.

```ruby

> case 1

Reviewer: ```suggestion
case value
```

when 1
  "One"
when 2
  "Two"
else
  "Other"
end
# => "One"

> # This is the same as:
> tmp = 1
> if 1 === tmp
>   "One"
> elsif 1 === tmp
>   "Two"
> else
>   "Other"
> end

Reviewer: ```suggestion
# This is the same as:
value = 1
if 1 === value
  "One"
elsif 1 === value
  "Two"
else
  "Other"
end
```

Reviewer: I would argue that the two code examples are not the same though, where 2 is explicitly matched, while the if statement will never get to the `else`.

Author: It is a type on line 30 if it is that you are referring to?

Reviewer: Oh, a "typographical" error, yeah, I believe so.  I know that it is a logic error as well, though, did not identify that it might be a typo.
